### PR TITLE
BUG/ENH: rework maximum flow preprocessing

### LINKED ISSUE #14385: csgraph.maximum_flow can cause Python crash for large but very sparse matrices
While reviewing #14358, I ran into the following issue (which affects both the implementation in that PR and the currently available SciPy implementation):

#### Reproducing code example:

Consider the flows calculation of the following very simple graph (in which each vertex has exactly one successor), represented by the matrix with ones on the superdiagonal and zeros elsewhere:

```python
import numpy as np
from scipy.sparse import csr_matrix
from scipy.sparse.csgraph import maximum_flow

def calc_flow(n):
    indices = np.arange(1, n)
    indptr = np.array(list(range(n)) + [n - 1])
    data = np.ones(n - 1, dtype=int)
    A = csr_matrix((data, indices, indptr), shape=(n, n))
    return maximum_flow(A, 0, n - 1)
```

On my machine, running this for `n in (10, 100, 1_000, 10_000)`, I get the expect result (a flow of 1), but for `n = 100_000`, Python crashes. This seems to suggest faulty memory allocation somewhere in the preprocessing for the flow calculation; the graph itself is so small that performing the calculation should be unproblematic.

#### Scipy/Numpy/Python version information:

```
1.6.1 1.19.2 sys.version_info(major=3, minor=7, micro=6, releaselevel='final', serial=0)
```

### PR DESCRIPTION
**Pull request by fuglede**
#### Reference issue
Closes gh-14385

#### What does this implement/fix?
The original preprocessing for `scipy.sparse.csgraph.maximum_flow` had an implicit assumption that the input graph had fewer than 2^16 vertices. When a larger graph was used, the preprocessing would cause an overflow that could crash the interpreter or, worst case, produce incorrect results.

While it would be straightforward enough to bump all integer sizes to allow larger graphs, here we take the approach of rewriting the preprocessing, following the lead of @mhwende who submitted an improvement [to the mailing list](https://mail.python.org/pipermail/scipy-dev/2020-January/023914.html) last year. Roughly, the change amounts to rewriting the preprocessing from NumPy natives to Cython. We do not include the DFS version of Edmonds--Karp also included in @mhwende's patch, but that would be a natural follow-up PR.

In addition to fixing the issue, this also comes with some very nice performance improvements (which is really the reason for preferring this approach over a quick fix); see the original mailing list post and [this comment](https://github.com/scipy/scipy/pull/14358#issuecomment-878210839) for benchmarks indicating what kind of improvements to expect. This becomes particularly useful in the context of gh-14358: After the approval of that PR, the preprocessing becomes a significant part of the overall runtime of the solvers, and the improvement here becomes a significant relative speed-up. After gh-14358 is completed, it would also make sense to include allow `'ford_fulkerson_dfs'` as an accepted value for the then-introduced `method`-parameter.

Moreover, the code is simply easier to follow than the original NumPy implementation.

#### Benchmark

Here, we compare the result for both available max flow solvers on a collection of different inputs ([more details here](https://github.com/scipy/scipy/pull/14358#issuecomment-878210839)). Here, "fast preprocessing" refers to the context of this pull request, so in the plot below, the first and third groups of bars are "without this PR", and the second and fourth groups are "after this PR".

image

| Type                  | Method       |       Time | Fast preprocessing   |
|-----------------------|--------------|------------|----------------------|
| Random (density 0.02) | dinic        | 0.00282028 | True                 |
| Random (density 0.02) | edmonds_karp | 0.00346423 | True                 |
| Random (density 0.1)  | dinic        | 0.0107675  | True                 |
| Random (density 0.1)  | edmonds_karp | 0.0199929  | True                 |
| Random (density 0.5)  | dinic        | 0.0435537  | True                 |
| Random (density 0.5)  | edmonds_karp | 0.284993   | True                 |
| Random (density 0.9)  | dinic        | 0.0517703  | True                 |
| Random (density 0.9)  | edmonds_karp | 0.842765   | True                 |
| Layered [500]*4       | dinic        | 0.0372847  | True                 |
| Layered [500]*4       | edmonds_karp | 0.50606    | True                 |
| Random (density 0.02) | dinic        | 0.0109687  | False                |
| Random (density 0.02) | edmonds_karp | 0.00847123 | False                |
| Random (density 0.1)  | dinic        | 0.0420544  | False                |
| Random (density 0.1)  | edmonds_karp | 0.0517822  | False                |
| Random (density 0.5)  | dinic        | 0.169735   | False                |
| Random (density 0.5)  | edmonds_karp | 0.409849   | False                |
| Random (density 0.9)  | dinic        | 0.206571   | False                |
| Random (density 0.9)  | edmonds_karp | 0.96701    | False                |
| Layered [500]*4       | dinic        | 0.198336   | False                |
| Layered [500]*4       | edmonds_karp | 0.626953   | False                |

#### Source code
<details>
Save the below as `benchmark.py`:

```python
import timeit
import numpy as np
from scipy.sparse import csr_matrix, rand
from scipy.sparse.csgraph import maximum_flow


def layered_graph():
    # Adopted fromhttps://mail.python.org/pipermail/scipy-dev/2020-January/023914.html
    layer_sizes = np.repeat(500, 4)
    num_layers = layer_sizes.shape[0]
    num_nodes = 2 + np.sum(layer_sizes)
    num_edges = layer_sizes[0] + layer_sizes[-1]
    for i in range(layer_sizes.size - 1):
        num_edges += layer_sizes[i] * layer_sizes[i + 1]
    np.random.seed(42)
    data = np.random.randint(0, 2, num_edges)
    indices = np.zeros(num_edges, dtype=int)
    indptr = np.zeros(num_nodes + 1, dtype=int)
    ptr = 0
    for j in range(layer_sizes[0]):
        indices[ptr] = 2 + j
        ptr += 1
    indptr[1] = ptr
    indptr[2] = ptr
    layer_sum = 0
    for k in range(num_layers - 1):
        next_layer_sum = layer_sum + layer_sizes[k]
        for i in range(layer_sizes[k]):
            for j in range(layer_sizes[k + 1]):
                indices[ptr] = 2 + next_layer_sum + j
                ptr += 1
            indptr[2 + layer_sum + i + 1] = ptr
        layer_sum = next_layer_sum
    for i in range(layer_sizes[-1]):
        indices[ptr] = 1
        ptr += 1
        indptr[2 + layer_sum + i + 1] = ptr
    indptr[num_nodes] = ptr
    a = csr_matrix((data, indices, indptr), shape = (num_nodes, num_nodes))
    return a


def make_cases():
    for density in (0.02, 0.1, 0.5, 0.9):
        graph = (rand(1000, 1000, density, random_state=42, format='csr')*100).astype(np.int32)
        yield (f'Random (density {density})', graph, 999)
    yield ('Layered [500]*4', layered_graph(), 1)


print('Type,Method,Time')
for graph_type, graph, sink in make_cases():
    for method in ('dinic', 'edmonds_karp'):
        t = timeit.Timer(lambda: maximum_flow(graph, 0, sink, method=method))
        mean_time = np.mean(t.repeat(10, 1))
        print(f'{graph_type},{method},{mean_time}')
```

To run this on after and before, from the SciPy root at current `master`, we run:

```
$ python benchmark.py > new.csv
$ git checkout 915e6d3c1b36e22b3d62bc6144fdb01066dd00ce
$ python setup.py build_ext --install
$ python benchmark.py > old.csv
```

Finally, to combine the two benchmarks and generate the outputs, we run the following:

```python
import pandas as pd
import tabulate


df_new = pd.read_csv('new.csv')
df_new['Fast preprocessing'] = True
df_old = pd.read_csv('old.csv')
df_old['Fast preprocessing'] = False
df = pd.concat([df_new, df_old])

print(tabulate.tabulate(df.values, headers=df.columns, tablefmt='github'))
df.groupby(['Type', 'Method', 'Fast preprocessing']).Time.mean().unstack(0).plot.bar()
```

</details>

cc @tupui @czgdp1807 @mhwende

### CODE REVIEW

**Comment by czgdp1807** [comment]
> We do not include the DFS version of Edmonds--Karp also included in @mhwende's patch, but that would be a natural follow-up PR.

I will be happy to take this up, once this and Dinic algo PR go in.



> After [gh-14358](https://github.com/scipy/scipy/pull/14358) is completed, it would also make sense to include allow `'ford_fulkerson_dfs'` as an accepted value for the then-introduced `method`-parameter.

This too. 

Let me know if that's possible. Thanks.

File: scipy/sparse/csgraph/_flow.pyx [hunk 1]
```diff
@@ -208,27 +208,39 @@ def maximum_flow(csgraph, source, sink):
...
+
     residual_array = np.asarray(residual)
     residual_matrix = csr_matrix((residual_array, m.indices, m.indptr),
                                  shape=m.shape)
     source_flow = residual_array[m.indptr[source]:m.indptr[source + 1]]
     return MaximumFlowResult(source_flow.sum(), residual_matrix)
 
 
+@cython.boundscheck(False)
+@cython.wraparound(False)
```
Comment (czgdp1807) [inline scipy/sparse/csgraph/_flow.pyx:243]: N.B. - https://github.com/scipy/scipy/pull/14358#discussion_r664717068

**Comment by fuglede** [comment]
> I will be happy to take this up, once this and Dinic algo PR go in.

Sounds good; it'd really just amount to [these few lines of code](https://github.com/mhwende/scipy/blob/e7be2dc29364fe95f6fae20a4c0775716cdba5e5/scipy/sparse/csgraph/_flow.pyx#L428-L433).

I haven't seen a case, however, where that approach is actually better than your implementation of Dinic's algorithm though. The [original mailing list post](https://mail.python.org/pipermail/scipy-dev/2020-January/023914.html) is concerned with the task of finding the number of disjoint paths in a layered graph, but Dinic's algorithm behaves very well on those (the layers corresponding to the levels); moreover, if one really wants to specialize to that example, chances are that the layers themselves are given, so the BFS step of Dinic can be skipped entirely. Moreover, Ford--Fulkerson with DFS search has exponential worst case complexity, but use cases could exist; it would just be useful to have them before adding an additional method.

**Comment by fuglede** [comment]
And by the way, @czgdp1807, if you're keen to work on these algorithms, I think the missing piece of the puzzle between `maximum_flow` and `min_weight_full_bipartite_matching` would be an efficient solver for the [minimum-cost flow problem](https://en.wikipedia.org/wiki/Minimum-cost_flow_problem), or one of its generalizations (perhaps [minimum-cost multi-commodity flow](https://en.wikipedia.org/wiki/Multi-commodity_flow_problem#Corresponding_optimization_problems)).

**Comment by rgommers** [comment]
This looks great. It did accumulate a merge conflict, could you resolve that @fuglede?

> In addition to fixing the issue, this also comes with some very nice performance improvements 

Would you mind adding a benchmark result to the PR description?

**Comment by czgdp1807** [comment]
>  I think the missing piece of the puzzle between `maximum_flow` and `min_weight_full_bipartite_matching` would be an efficient solver for the [minimum-cost flow problem](https://en.wikipedia.org/wiki/Minimum-cost_flow_problem), or one of its generalizations (perhaps [minimum-cost multi-commodity flow](https://en.wikipedia.org/wiki/Multi-commodity_flow_problem#Corresponding_optimization_problems)).

Sure. May be we can implement both (the best algorithms for specific and generalised versions). There are very few open source implementations for multi-commodity flow problems. For minimum cost flow I found [this](https://github.com/frangio68/Min-Cost-Flow-Class). Though its license doesn't allow changing the code. May be we can implement the network simplex algorithm for minimum cost flow from scratch. I am not sure how hard that would be. Let me know.

**Comment by fuglede** [comment]
@rgommers wrote:

> This looks great. It did accumulate a merge conflict, could you resolve that @fuglede?

Yep, done. After the resolution, the regression test added here segfaults with `method='dinic'` (but not `'edmonds_karp'`), so that needs to be resolved before this is done.

> Would you mind adding a benchmark result to the PR description?

Yes, let me just add the benchmarks from the other PR here directly.

@czgdp1807 wrote:

> There are very few open source implementations for multi-commodity flow problems.

So it seems. I can find a few, but it's a bit more niche than min cost flow, so maybe a better fit for something like NetworkX (although that's pure-Python, so we don't get the wonderful performance that the CSR/Cython implementations tend to get).

> For minimum cost flow I found this. 

There are a handful of Python implementations; a Python extension using the one you've found, there's a Python implementation in NetworkX, and then it's useful to compare with OR-Tools as well. I'd start by trying out what's out there, see what works, and whether it's possible to improve on the existing solvers while sticking to known non-exotic algorithms (cf. an earlier comment by @rgommers that it's best to stick to well-known algorithms for maintainability).

**Comment by fuglede** [comment]
The following example fails on `master`:

```python
    n = 5062
    indices = np.arange(1, n)
    indptr = np.array(list(range(n)) + [n - 1])
    data = np.ones(n - 1, dtype=np.int32)
    graph = csr_matrix((data, indices, indptr), shape=(n, n))
    res = maximum_flow(graph, 0, n - 1)
```

Perhaps the recursion is wrecking the stack? Will have to see; I tried the patch in [this comment](https://github.com/scipy/scipy/pull/14358#discussion_r669351157) which uses an ordinary stack but run into an infinite loop. Will dig a bit further. Feel free to chip in if you have a better approach @czgdp1807.

**Comment by czgdp1807** [comment]
IMHO, a minimum of 1.25e11 computation steps (assuming |E| ~ O(|V|)) will be required to find the maximum flow here (O(|V|^3)) (|V| = 5000). On an average 1e8 steps happen in 1s (at least on codeforces server). So more than 1000s will be required by an iterative solution if everything goes well. I will try to debug it if it's not the case.

**Comment by fuglede** [comment]
@czgdp1807 Regarding the infinite loop, you mean? One will run into that even if `n = 5`. The graph is just a straight line, so only one augmentation is required. Since the only difference between the augmentation steps in `_edmonds_karp` and `_dinic` is the use of BFS vs. DFS, and the additional check on `levels`, it's rather tempting to combine the two.

**Comment by czgdp1807** [comment]
> One will run into that even if `n = 5`. The graph is just a straight line, so only one augmentation is required

Ah! Okay. I will see. The problem is with the depth of the recursion call stack. The implementation in itself is correct I believe. To turn it into a correctly working iterative solution one might need to immitate that behaviour manually using stacks which might be buggy. I can try fixing the patch tomorrow.

**Comment by czgdp1807** [comment]
The iterative solution in https://github.com/scipy/scipy/pull/14358#discussion_r669351157 is working fine for me.
Following result is produced,
```
MaximumFlowResult with value of 1
```

**Comment by fuglede** [comment]
1 is the correct result, yes; the graph is just

```
X --1--> X --1--> X --1--> X -- ... --> X
```

But great that it works; I must have messed up merging the patch into `master` then.

I think the iterative approach is reasonably readable. In the corresponding code in `_edmonds_karp`, we avoid `progress` altogether, and keep the stack one-dimensional (and the cost of having to carry around `pred_edge`.

One thing that's not immediately clear to me (in the recursive or iterative approaches) is if we properly avoid visiting a vertex more than once. Doing so won't change correctness but is redundant, and is avoided through the check `pred_edge[t] == -1` in `_edmonds_karp`.

**Comment by czgdp1807** [comment]
May be you can try the new one. The older one was outdated and hence doesn't go well with the current code. Let me know if this one works. I will clean it up and reshare here so that it can be applied to your PR. I can open a new PR but may be that would be a bit redundant.

**Comment by fuglede** [comment]
@czgdp1807 I tried the new one and can confirm it works on the given example. I can also confirm your initial conclusion that there is no significant difference in performance.

If, however, I run it on a large number of examples, I eventually get different results between the solvers, unless I've managed to mess up the merge again; for example, the following has a non-empty output (where `dinic2` is the iterative version):

```python
for i in range(1000):
    graph = (rand(500, 500, format='csr', random_state=i) * 1000).astype(np.int32)
    ek = maximum_flow(graph, 0, 499, method='edmonds_karp').flow_value
    din1 = maximum_flow(graph, 0, 499, method='dinic').flow_value
    din2 = maximum_flow(graph, 0, 499, method='dinic2').flow_value
    if len({ek, din1, din2}) != 1:
        print(f'i={i}, ek={ek}, din1={din1}, din2={din2}')
        break
```
prints
```
i=24, ek=3819, din1=3819, din2=3854
```

**Comment by czgdp1807** [comment]
> In any case, we're getting there. I agree that opening a new PR would be a bit much, but it would make my workflow a bit simpler if you include your changes on a separate branch on your fork, so that would be appreciated.

No problem. I will open a new PR once I correctly implement the iterative solution.

**Comment by czgdp1807** [comment]
> there's a Python implementation in NetworkX

I am able to find it [here](https://github.com/networkx/networkx/blob/main/networkx/algorithms/flow/mincost.py). May be we can port it to Cython and compare it with OR-Tools implementation [here](https://developers.google.com/optimization/reference/graph/min_cost_flow/MinCostFlow?hl=en)? Let me know what you think.

**Comment by fuglede** [comment]
With #14422 merged, this one is now good to go as well.

> I am able to find it here. May be we can port it to Cython and compare it with OR-Tools implementation here?

I'd certainly be curious to see how it would perform, at least. The cost-scaling/push-relabel algorithms could be useful as well. Ultimately, someone else will have to decide what's an appropriate threshold for acceptable code complexity, but if the worst case is having another good solver in the Python ecosystem, then that's not too bad either. But yes, certainly a more involved task than the one's we're looking at here.

**Comment by fuglede** [comment]
CI failure is unrelated:

> scipy/stats/_qmc.py:891: error: Argument 2 to "tile" has incompatible type "Tuple[Union[int, integer[Any]], int]"; expected "Union[int, Sequence[int]]"  [arg-type]

**Comment by rgommers** [comment]
> I ran some benchmarks on the new method with the Cython-based preprocessing mentioned [in this mailing list thread](https://mail.python.org/pipermail/scipy-dev/2020-January/023914.html),

Would you mind adding the latest version of this benchmarking code in the PR description here (you can enclose it in `<details> ... </details>` if it's long). The mailing list version doesn't generate the image in this PR, and it'll likely be useful to have in the future.

> With #14422 merged, this one is now good to go as well.

Great! @czgdp1807 if this looks good to you too, I'll merge it.

**Comment by czgdp1807** [comment]
LGTM. The existing tests pass without change so everything should be good.

**Review by rgommers — APPROVED** [review]
Okay, in it goes then. Thanks @fuglede and @czgdp1807!

**Comment by fuglede** [comment]
Thanks both.

> Would you mind adding the latest version of this benchmarking code in the PR description here 

Done, and reran the benchmark with a few modifications.

**Comment by czgdp1807** [comment]
Starting work on https://github.com/scipy/scipy/pull/14392#issuecomment-882912248
Will open a PR and share the benchmarks and results there. Thanks.

**Pull request merged**